### DISCUSSION TITLE
Unable to remove default value from auth_backends

### DISCUSSION THREAD

**Discussion by Shubham3009** [apache/airflow#26084] [Q&A]
Hi ,
I am using Airflow 2.3.2 and getting below Warning while calling Airflow APIs ,

**airflow jobs check**
/opt/bitnami/airflow/venv/lib/python3.8/site-packages/airflow/configuration.py:356: FutureWarning: The auth_backends setting in [api] has had airflow.api.auth.backend.session added in the running config, which is needed by the UI. Please update your config before Apache Airflow 3.0.  warnings.warn(
Found one alive job.

 **airflow version**
/opt/bitnami/airflow/venv/lib/python3.8/site-packages/airflow/configuration.py:356: FutureWarning: The auth_backends setting in [api] has had airflow.api.auth.backend.session added in the running config, which is needed by the UI. Please update your config before Apache Airflow 3.0.  warnings.warn(
2.3.2

I tried to override AUTH_BACKENDS as below , but default values is still in use and not getting overridden by my value.
 - name: AIRFLOW__API__AUTH_BACKENDS
    value: "airflow.api.auth.backend.basic_auth" 

In airflow.cfg , Currently I see two values for auth_backends as below ,
auth_backends : airflow.api.auth.backend.basic_auth,airflow.api.auth.backend.session
-- | --

Because of this behavior , am unable to apply livness probe on SCHEDULER to autorestart if heartbeat not received. 

Please help to resolve this warning or override auth_backend default value.

**Comment by potiuk** [comment]
You likely have a problem with your configuration and the configuration you are looking at is not what you think it is.

If you have airflow.api.auth.backend.session in your configuration, then this warning should not be there.

I think the bitnami chart you or image or probe you are using has some erroneous behavour and does not properly apply the configuration and you should raise the issue to bitnami. This problem does not exist if you use the official Airflow Image  https://airflow.apache.org/docs/docker-stack/index.html. and official charat https://airflow.apache.org/docs/helm-chart/stable/index.html so my recommendation to solve the problem is to switch from Bitnami to ours.

Unfortunatley we cannot help when 3rd-party is makins some mistakes, you shoudl really raise the problem to them if you want to stick with them.

This is the main reason why we have our official image and chart that we can help our ousers and provide them with better support, if you choose to go with bitnami, then they should do any support of this kind.

Sorry for not being helpful with bitnami, but we do any help here in a free time and we have very little of that, Airflow is delivered for free and if a 3rd-party decides to release something on top of what we release - they take on the responsibiity of helping their users.

**Comment by binary-signal** [comment]
Running Airflow with the official image via docker-compose as provided from the documentation, having the same issue as @Shubham3009 

https://airflow.apache.org/docs/apache-airflow/stable/howto/docker-compose/index.html

On the docker-compose file I have exposed the environment variable
AIRFLOW__API__AUTH_BACKENDS: 'airflow.api.auth.backend.basic_auth'

Strangely Airflow webui displays 2 values for auth backends
![Screen Shot 2022-09-23 at 5 09 06 PM](https://user-images.githubusercontent.com/3959074/191993049-37038a5d-afb7-4fe3-abb1-5837a92ea92b.jpg)


![Screen Shot 2022-09-23 at 5 11 44 PM](https://user-images.githubusercontent.com/3959074/191993694-f2e428f4-a52c-41ba-a5b6-11754b35979a.jpg)

From inside the scheduler container also 2 values are been reported  when running the following command:
```shell
default@fb5ec71f2dcb:/opt/airflow$ airflow config get-value api auth_backends

/home/airflow/.local/lib/python3.10/site-packages/airflow/configuration.py:363: FutureWarning: The auth_backends setting in [api] has had airflow.api.auth.backend.session added in the running config, which is needed by the UI. Please update your config before Apache Airflow 3.0.
  warnings.warn(

airflow.api.auth.backend.basic_auth,airflow.api.auth.backend.session

....
```

Tested on Airflow versions 2.3.3, 2.3.4, 2.4.0

**Reply by potiuk** [reply]
I think you are not exposing it correctly in the docker-compose. Note that there are YAML anchors in the compose and you need to understand exactly what you are doing when you are modifying the docker-compose. It's not meant to be updated or extended and if you are doing it, then likely you might be mislead on how docker compose sets the variables.

Basically I think you shoudl not realy on "docker-compose" of ours being anything else by a quick-start guide. You likely made a mistake somewhere. Similarly as in https://github.com/apache/airflow/discussions/26646 for example. 

But what it is, I cannot help - I had been burned far too many times by docker-compose misleading behaviour and I recommend everyone to use Helm Chart and K8S - which is the only way we support and recommend in the community. If you are choosing the Docker-Compoe, you likely chose a rough road.

**Reply by binary-signal** [reply]
A quick work around was to expose explicitly 2 auth backends in order to make the deprecation  warning go away in docker-compose file.

```shell
  AIRFLOW__API__AUTH_BACKENDS: "airflow.api.auth.backend.basic_auth,airflow.api.auth.backend.session"
```


From what I understood airflow  "injects" the session backend if it's missing from the config/env during runtime, in order to avoid users getting locked out from the UI, if they upgrade to a future version which reasonable.

I am not sure if this is also related to a quite recent change in airflow 2.3.0 which allows comma separated values for  auth backends and the parsing logic related to it

Snippet from the configuration reference https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html
....
auth_backends

New in version 2.3.0.

Comma separated list of auth backends to authenticate users of the API. See https://airflow.apache.org/docs/apache-airflow/stable/security/api.html for possible values. (“airflow.api.auth.backend.default” allows all requests for historic reasons)
.....



Alternatively an other  solution is to  replace the default config value for authentication backends in airflow.cfg

from:
```shell
# Comma separated list of auth backends to authenticate users of the API. See
# https://airflow.apache.org/docs/apache-airflow/stable/security/api.html for possible values.
# ("airflow.api.auth.backend.default" allows all requests for historic reasons)
auth_backends = airflow.api.auth.backend.session
```

to:
```shell
# Comma separated list of auth backends to authenticate users of the API. See
# https://airflow.apache.org/docs/apache-airflow/stable/security/api.html for possible values.
# ("airflow.api.auth.backend.default" allows all requests for historic reasons)
auth_backends = airflow.api.auth.backend.basic_auth,airflow.api.auth.backend.session
```

**Reply by potiuk** [reply]
I think you just had different configuration in webserver and scheduler (either via file or variable). You should make sure that all the configuration is set the same way for all components. See:

https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html

> Use the same configuration across all the Airflow components. While each component does not require all, some configurations need to be same otherwise they would not work as expected. A good example for that is [secret_key](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#config-webserver-secret-key) which should be same on the Webserver and Worker to allow Webserver to fetch logs from Worker.

**Comment by jli113** [comment]
Same on ubuntu 22.04 with " curl -LfO 'https://airflow.apache.org/docs/apache-airflow/2.5.0/docker-compose.yaml' "

**Reply by jli113** [reply]
Turns out it was not a problem, I changed auth_backends = airflow.api.auth.backend.basic_auth to auth_backends = airflow.api.auth.backend.session. Then, I followed the steps in the [guide](https://airflow.apache.org/docs/apache-airflow/stable/howto/docker-compose/index.html)
